### PR TITLE
feat: add optional column for adapters

### LINKED ISSUE #356: Parallel Analysis of same Technology
## Feature Request

Add a flag for parallel data analysis, which was created with different sequencing kits but on the sample technology.

- [ ] Discuss how to implement this. Config? Sample Sheet?

## Motivation

For the paper, we need a report (for showcasing) that contains sample sequenced Illumina with Nimagen primers, Illumina with Revelo primers, Illumina with Illumina primers and ONT with ARTIC kit.

## Alternatives

None

## Additional context

None

### PR DESCRIPTION
**Pull request by thomasbtf**
# Description

This PR adds the option to specify adapters sequence for trimming via the sample sheet. However, this is just an option. Per default the adapates specified in the config are used. Only for samples where the column `adapters` in the sample sheet these "custom adapters" are used.

It also provides some "predefined" updates sequences, which can be used by speceifing `revelo-rna-seq` or `nimagen-easy-seq` in the sample sheet.

Lastly, the docs are updated.

## Related Issue

closes #356 

## Checklist

- [X] I've formatted the PR title in accordance with the [structure of the conventional commit messages](https://www.conventionalcommits.org/en/v1.0.0/).
- [X] I've updated the code style using [`pre-commit`](https://ikim-essen.github.io/uncovar/dev-guide/contributing/#pre-commit) if needed.
- [X] I've updated or supplemented the documentation as needed.
- [X] I've read the [`CODE_OF_CONDUCT.md`] document.
- [X] I've read the [`CONTRIBUTING.md`] guide.

[`CODE_OF_CONDUCT.md`]: https://github.com/IKIM-Essen/uncovar/blob/master/CODE_OF_CONDUCT.md
[`CONTRIBUTING.md`]: https://github.com/IKIM-Essen/uncovar/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: docs/user-guide/configuration.md [hunk 1]
```diff
@@ -1 +1,63 @@
 # Configuration
+
+## Adapters
+
+There are three ways to transfer adapter sequences to UnCoVar to remove them
+from the raw data.
```
Comment (alethomas) [inline docs/user-guide/configuration.md:6]:
```suggestion
There are three ways to transfer adapter sequences to UnCoVar for removing them
from the raw data.
```

File: docs/user-guide/configuration.md [hunk 2]
```diff
@@ -1 +1,63 @@
...
+
+### Config File
+
+The adapter sequences used can be entered in the config file under
```
Comment (alethomas) [inline docs/user-guide/configuration.md:10]:
```suggestion
The adapter sequences used can be specified in the config file under
```

File: docs/user-guide/configuration.md [hunk 3]
```diff
@@ -1 +1,63 @@
...
+`preprocessing` -> `kit adapters`.
+
+For **paired-end data**, the adapters can be detected by per-read overlap
+analysis, which seeks the overlap of each pair of reads. The adapter sequences
```
Comment (alethomas) [inline docs/user-guide/configuration.md:14]:
```suggestion
analysis, which seeks the overlap for each pair of reads. The adapter sequences
```

File: docs/user-guide/configuration.md [hunk 4]
```diff
@@ -1 +1,63 @@
...
+can specify the adapter sequences for read one by `—adapter_sequence` and for
```
Comment (alethomas) [inline docs/user-guide/configuration.md:15]:
```suggestion
can be specified for read one by `—adapter_sequence` and for
```

File: docs/user-guide/configuration.md [hunk 5]
```diff
@@ -1 +1,63 @@
...
+read two by`—adapter_sequence_r2`. An example is for [Illuminas TruSeq library] (<https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/truseq-rna-v2.html>)
```
Comment (alethomas) [inline docs/user-guide/configuration.md:16]:
```suggestion
read two by`—adapter_sequence_r2`. An example for [Illuminas TruSeq library] (<https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/truseq-rna-v2.html>)
```

File: docs/user-guide/configuration.md [hunk 6]
```diff
@@ -1 +1,63 @@
...
+`—adapter_sequence` option.
+
+```yaml
+"--adapter_sequence = AGATCGGAAGAGCACACGTCTGAACTCCAGTCA"
+```
+
+### Sample Sheet
+
+The second way to remove adapter sequences is to specify the adapter sequence
+directly sample in the sample sheet. The adapters must enter it in a column
```
Comment (alethomas) [inline docs/user-guide/configuration.md:34]:
```suggestion
The second way to remove adapter sequences is to specify the adapter sequence
 per sample in the sample sheet. The adapters must be entered in a column
```

File: docs/user-guide/configuration.md [hunk 7]
```diff
@@ -1 +1,63 @@
...
+an exemplary samples sheet:
+
+| sample_name | fq1         | fq2         | date       | is_amplicon_data | technology | adapters                                           |
+|-------------|-------------|-------------|------------|------------------|------------|----------------------------------------------------|
+| example-1   | PATH/TO/fq1 | PATH/TO/fq2 | 1970-01-01 | 1                | illumina   | --adapter_sequence=ACGT --adapter_sequence_r2=TGCA |
+| example-2   | PATH/TO/fq  |             | 1970-01-01 | 1                | ion        | --adapter_sequence=ACGT                            |
+
+If an adapter sequence is entered into the sample sheet for one sample, this
+adapter sequence is obviously used to trim the sequences of this sample. For
+empty entries, UnCoVar uses the adapter sequence from the config file.
```
Comment (alethomas) [inline docs/user-guide/configuration.md:45]:
```suggestion
If an adapter sequence is specified for a sample in the sample sheet, this
adapter sequence is used to trim the sequences of only this sample. For
empty entries, UnCoVar uses the adapter sequence from the config file.
```

File: docs/user-guide/configuration.md [hunk 8]
```diff
@@ -1 +1,63 @@
...
+
+### Pre-Defined Adapters
+
+UnCoVar supports two different sequencing kits and their respective adapters,
+namely:
+
+1. [Revelo RNA-Seq library preparation kit](https://lifesciences.tecan.com/revelo-rna-seq-library-prep-kit?p=tab--5)
+1. [EasySeq RC-PCR SARS CoV-2  Whole Genome Sequencing kit](https://www.nimagen.com/shop/products/rc-cov096/easyseq-sars-cov-2-novel-coronavirus-whole-genome-sequencing-kit)
+
+The `adapters` column in the sample sheet is used to trim the adapters sequences
```
Comment (alethomas) [inline docs/user-guide/configuration.md:55]:
```suggestion
The `adapters` column in the sample sheet is used to trim the adapter sequences
```

File: workflow/rules/common.smk [hunk 9]
```diff
@@ -826,39 +826,35 @@ def get_depth_input(wildcards):
...
+        samples.dropna(subset=["adapters"], inplace=True)
+        adapters = samples.loc[wildcards.sample]["adapters"]
+
+        # predefined adapters
+        # https://lifesciences.tecan.com/revelo-rna-seq-library-prep-kit
+        if adapters == "revelo-rna-seq":
+            return "--adapter_sequence AGATCGGAAGAGCACACGTCTGAACTCCAGTCA --adapter_sequence_r2 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"
+        # https://www.nimagen.com/shop/products/rc-cov096/easyseq-sars-cov-2-novel-coronavirus-whole-genome-sequencing-kit
+        elif adapters == "nimagen-easy-seq":
+            return "--adapter_sequence GCGAATTTCGACGATCGTTGCATTAACTCGCGAA --adapter_sequence_r2 AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"
```
Comment (alethomas) [inline workflow/rules/common.smk:852]: I'd still prefer to keep these in the config and not hard coded in the common.smk. Or am I missing something here?

**Review by alethomas — CHANGES_REQUESTED** [review]

File: workflow/rules/common.smk (hunk 9 above)
Comment (thomasbtf) [reply workflow/rules/common.smk:852]: These hardcoded sequences are just used if you enter "revelo-rna-seq" or "nimagen-easy-seq" in the new adapter column. Else the sequences in the config file are used.

However, for generating the ref data set report, we will have to use this new column.

**Review by alethomas — APPROVED** [review]

**Pull request merged**